### PR TITLE
refactor: Update SRP helpers

### PR DESCRIPTION
**Pull request by cshfang**
#### Description of changes

This PR completes a refactoring of the Auth SRP helpers that appeared to be partially underway. This resulted in duplication of code and inconsistency in behavior. Following changes were made:

* Extracted common function to improve testability and extensibility. This is especially useful for the `*.native` code split so we do not need to override prototypes for the `AuthenticationHelper` class.
* Completed the functional refactoring of non-instance specific helpers to reduce code duplication. This resulted in a slight bundle size improvement but could also have memory improvements in runtime.
* Promisified APIs built on top of the `BigInteger`'s `modPow` function which expects callbacks. This removes the need for a callback promisifier util during testing and improves code readability. 
* Reorganized `AuthenticationHelper` to expose an asynchronous construction function to remove some race conditions in the class logic.

#### Description of how you validated changes
* Tested locally with a SRP login
* `yarn test`

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by israx — DISMISSED** [review]
Thanks @cshfang  for putting these together. I was able to test this code when cognito returns  `DEVICE_SRP` and `DEVICE_PASSWORD_VERIFIER` and it is working fine!

**Review by jimblanc — DISMISSED** [review]
Looks great, thanks for doing this

**Comment by cshfang** [comment]
> Thanks @cshfang for putting these together. I was able to test this code when cognito returns `DEVICE_SRP` and `DEVICE_PASSWORD_VERIFIER` and it is working fine!

Awesome @israx thank you so much for testing

File: packages/auth/__tests__/providers/cognito/utils/srp/calculate/calculateS.test.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+				x,
+				B,
+				N,
+				U,
+			})
+		).rejects.toThrow();
+	});
+});
+
+// 	it('modPow throws an error ', async () => {
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/auth/__tests__/providers/cognito/utils/srp/calculate/calculateS.test.ts:78]: Oops yep. Copied existing tests over as reference to make sure I covered those cases at a minimum and forgot to remove

**Review by israx — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Pull request merged**